Firestore/core/src/firebase/firestore/util/executor_std.h

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EXECUTOR_STD_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EXECUTOR_STD_H_
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>  // NOLINT(build/c++11)
+#include <deque>
+#include <mutex>   // NOLINT(build/c++11)
+#include <string>
+#include <thread>  // NOLINT(build/c++11)
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/executor.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+namespace internal {
+
+// A thread-safe class similar to a priority queue where the entries are
+// prioritized by the time for which they're scheduled. Entries scheduled for
+// the exact same time are prioritized in FIFO order.
+//
+// The main function of `Schedule` is `PopBlocking`, which sleeps until an entry
+// becomes available. It correctly handles entries being asynchonously added or
+// removed from the schedule.
+//
+// The details of time management are completely concealed within the class.
+// Once an entry is scheduled, there is no way to reschedule or even retrieve
+// the time.
+template <typename T, typename Duration = std::chrono::system_clock::duration>
+class Schedule {
+  // Internal invariants:
+  // - entries are always in sorted order, leftmost entry is always the most
+  //   due;
+  // - each operation modifying the queue notifies the condition variable `cv_`.
+ public:
+  // Entries are scheduled using absolute time.
+  using TimePoint =
+      std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+  // Schedules an entry for the specified time due. `due` may be in the past.
+  void Push(const T& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{value, due});
+  }
+  void Push(T&& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{std::move(value), due});
+  }
+
+  // If the queue contains at least one entry for which the scheduled time is
+  // due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue, moves it into `out`, and returns true. If no
+  // entry is due, doesn't modify `out` and returns false. `out` may be
+  // `nullptr` (in which case the value will be simply discarded).
+  bool PopIfDue(T* const out) {
```

Comment:
Re leaving it in: no objection from me.

Re optional vs statusor: sgtm

Re: user provided mv/cp ctors that could throw: Ah, that make sense. Let me see if I've got this straight:

If it throws, we could be in a state where we've lost the operation since it's been removed from our queue, but not returned to the user (except in the form of an exception), whereas before, we'd keep it in the queue.

Loosing it seems bad, but keeping it in the queue also seems bad since presumably the next call to `pop` will fail with the same issue. Maybe if the error was correctable (eg allocation related), caller could correct it (eg free memory) and try again. But otherwise it seems like this situation is sufficiently bad, that it's unclear to me if there's anything the caller could reasonably do to recover.

Reply:
> If it throws, we could be in a state where we've lost the operation since it's been removed from our queue, but not returned to the user (except in the form of an exception), whereas before, we'd keep it in the queue.

Yes, exactly.

I can restore the previous pattern if necessary. @wilhuff Gil, what do you think?